Polymer2

mdgbayly opened the pull request:
Upgrade to Polymer 2/hybrid.

Should probably be a major version change to to the changes from `content` to `slot`.

It was a bit troublesome adding the support for the shadom/shady DOM/CSS properties as some of the Polymer.Settings that were controlling this do not seem to be available in Polymer 2.
So I had to remove the label that warns you where those things are not supported.
I guess I could have added conditional 1/2 code but was getting lazy...and trying to keep it simple.
So if you choose 'Use Shadow' and it's not supported it just refreshes the template, but 'Use Shadow' is still displayed.

One other difference which is interesting in Polymer 2 is that if you use the CSS shim option with shadow DOM then it basically breaks the page whereas in Polymer 1, it seems like if you enable shadow DOM it forces the shim to be off.

dbatiste commented on d2l-demo-template.html:
```diff
@@ -239,33 +241,42 @@ <h1 class="d2l-heading-4" style="margin: 0;">[[title]]</h1>
+						useShadow.setAttribute('href', href);
+
 					}
-				} else {
```
Interesting, this is relevant for browsers that don't support native shadow-DOM.  I am curious what is now available (with Polymer 2) in Polymer.Settings.  We may be able to do our own check for that support.

dbatiste commented:
@mdgbayly : maybe check for `shadowRoot`, as done in this [dom-helper](https://github.com/Brightspace/d2l-polymer-behaviors-ui/blob/master/d2l-dom.html#L32)?

mdgbayly commented:
i should probably close the PR until I have this working to stop spamming you as I'm testing stuff out by pushing to my branch rather than messing with bower link.

I think testing for shadowRoot would require setting up some kind of shadow DOM element. Not sure if that will be chicken and egg when we had shady DOM enabled...

This StackOverflow posts suggest using this for feature detection but that actually gives a false positive in FF as attachShadow seems to be defined:

https://stackoverflow.com/questions/29730398/how-to-check-if-a-browser-supports-shadow-dom


    if(document.head.createShadowRoot || document.head.attachShadow) {
        // I can shadow DOM
    } else {
        // I can't
    }

I'm going to park this for a while as I need to get back to some feature work but will keep working on it as a side task...

mdgbayly commented:
@dbatiste @dlockhart 

Updated PR. To make the shadow DOM feature detection work for Polymer 2, i'm just checking for the v0 createShadowRoot because attachShadow seemed to be giving false positives in FF and Safari. Maybe we can revisit that at some point but for now it seems to be correctly detecting in IE/Edge/FF/Chrome/iOS Safari.

Added in the hybrid linting stuff and updated the package.json to install the variants which also covers it for the travis install, although I don't think it's really being used currently by the travis build as that is just linting rather than running any tests.  

I think we will probably be OK if we just do a major version bump on this repo to account for the content/slot changes.  I don't think we will need to do a major version bump on `colors`.  I did some more testing locally using this version of d2l-demo-template-ui in a local colors repo, which I then bower linked to a 3rd polymer 1.0 project and it did not appear to have versioning issues with the d2l-demo-template-ui stuff.

Not entirely sure if that will be the case when it's actually pulling from bower rather than using bower linking, but maybe we can try and get away with a minor version bump on colors, and if it causes issues retag with a major version?

I will revisit the PRs for the typography and colors changes now.
